DirectProgramming/C++SYCL_FPGA/ReferenceDesigns/convolution2d/include/vvp_stream_adapters.hpp

```diff
@@ -0,0 +1,347 @@
+/// hang.
+/// @tparam PixelPipe The pipe from which pixels will be read. This pipe's
+/// payload should be a `StreamingBeat` templated on a `DataBundle`, which is
+/// itself templated on a payload of type `PixelType`.
+/// @tparam PixelType The type that represents each pixel. This may be a scalar
+/// (such as an `int`) or a `struct` of 'plain old data'.
+/// @param q SYCL queue where your oneAPI kernel will run
+/// @param[in] rows Image rows (height)
+/// @param[in] cols Image columns (width)
+/// @param[out] out_img Pointer to place image pixels read from `PixelPipe`
+/// @param[out] sidebands_ok Indicates if sideband signals are correct
+/// @param[out] defective_frames Indicates how many defective frames were read
+/// before the current frame
+/// @return `false` if the packet reader ends in an undefined state
+template <typename PixelPipe, typename PixelType>
+bool ReadFrameFromPipe(sycl::queue q, int rows, int cols, PixelType *out_img,
+                       bool &sidebands_ok, int &defective_frames) {
+  ///////////////////////////////////////
+  // Extract parameters from PixelPipe
+  ///////////////////////////////////////
+
+  // the payload of PixelPipe should be a StreamingBeat
+  using StreamingBeatType = typename ExtractPipeType<PixelPipe>::value_type;
+
+  // the payload of PixelPipe should be a StreamingBeat, whose payload is a
+  // DataBundle
+  using DataBundleType = BeatPayload<PixelPipe>;
+  constexpr int kPixelsInParallel =
+      fpga_tools::ExtractDataBundleType<DataBundleType>::kBundlePayloadCount;
+  using PixelTypeCalc = typename fpga_tools::ExtractDataBundleType<
+      DataBundleType>::BundlePayloadT;
+
+  // sanity check
+  static_assert(std::is_same<PixelTypeCalc, PixelType>::value, "mismatched");
+  if (0 != (cols % kPixelsInParallel)) {
+    std::cerr << "ERROR: kPixelsInParallel must be a factor of cols!!";
+    return false;
+  }
+
+  ////////////////////////////////////////////////////////////
+  // Consume the beats from PixelPipe, and place into out_img
+  ////////////////////////////////////////////////////////////
+
+  std::cout << "INFO: Reading data from pipe with " << kPixelsInParallel
+            << " pixels in parallel. " << std::endl;
+
+  int eop_count = 0;
+  bool passed = true;
+  int i_base = 0;
+
+  while (eop_count < rows) {
+    // expect SOP at the beginning of each frame
+    bool sop_expected = (i_base == 0);
+    // expect EOP at the end of each line. This calculation is valid because we
+    // require `cols` to be a multiple of `kPixelsInParallel`.
+    bool eop_expected = (0 == ((i_base + kPixelsInParallel) % cols));
+    int empty_expected =
+        (eop_expected
+             ? ((kPixelsInParallel - ((rows * cols) % kPixelsInParallel)) %
+                kPixelsInParallel)
+             : 0);
+    bool sop_calc = false;
+    bool eop_calc = false;
+    int empty_calc = 0;
+    StreamingBeatType out_beat = PixelPipe::read(q);
+    DataBundleType outBundle;
```

Comment:
It would seem to be consistent outBundle would be normal snake_case out_bundle.

Reply:
Thanks, fixed this.